Making navbar + a few other sections more modular

choldgraf opened the pull request:
This is an attempt at making the navbar more modular, so that it can be more easily customized. It does a couple of things, though the end-result should look the exact same as what we have right now.

1. It defines two new configuration: `html_theme_options: {"navbar_left": ..., "navbar_menu": ...}`
   
   Each of these are a list of templates, similar to `html_sidebars`, that will populate the two sections of the navbar.
2. Removes the configuration value to put the search in the sidebar or the navbar, since people can do this via `html_theme_options: {"navbar_menu": ["search-field.html"]}` now.
3. Splits our navbar HTML into a few components in `_templates/`
4. Adds an `ethical ads` template
5. Tries to document a bit of this, and adds a little `version` template for the docs so that we can see how one can extend the left navbar area.

@jorisvandenbossche what do you think about this? If you're +1 then I can add some tests.

The docs previous shouldn't look any different, but just for reference: https://pydata-sphinx-theme--355.org.readthedocs.build/en/355/user_guide/index.html

(the one difference is that there should now be a little version next to the logo)

OriolAbril commented:
I have tested it locally and can confirm that the default look is not altered and I can use the `navbar-version.html` approach to add the version. Thanks!

MegChai commented:
 How about adding a version switcher instead of just display it in the top nav-bar? (#350 ) We can put a `version.json` file to control how it generates. https://github.com/MegEngine/pydata-sphinx-theme/pull/1

jorisvandenbossche commented on docs/conf.py:
```diff
@@ -21,9 +21,11 @@
 
-# The full version, including alpha/beta/rc tags
-release = "0.0.1dev0"
+from pathlib import Path
 
+for ii in Path("../pydata_sphinx_theme/__init__.py").read_text().split("\n"):
+    if ii.startswith("__version__"):
+        version = release = ii.split("= ")[-1].strip('"').replace("dev0", "")
```
You could also import the package and get `pydata_sphinx_theme.__version__` ?

jorisvandenbossche commented on docs/user_guide/configuring.rst:
```diff
@@ -289,6 +305,65 @@ use this pattern:
+   html_theme_options = {
+   ...
+   "navbar_left": ["navbar-logo.html"],
+   "navbar_menu": ["navbar-menu-nav.html", "navbar-menu-search.html", "navbar-menu-buttons.html"],
```
```suggestion
   "navbar_menu": ["navbar-menu-nav.html", "navbar-menu-buttons.html"],
```

?

jorisvandenbossche commented on docs/user_guide/configuring.rst:
```diff
@@ -289,6 +305,65 @@ use this pattern:
+.. note::
+
+   Generally speaking, navbar menu items will be **right-aligned**. However, the
+   ``navbar-menu-nav.html`` template any anything that comes before it will
```
```suggestion
   ``navbar-menu-nav.html`` template and anything that comes before it will
```

jorisvandenbossche commented on docs/user_guide/configuring.rst:
```diff
@@ -289,6 +305,65 @@ use this pattern:
+
+   Generally speaking, navbar menu items will be **right-aligned**. However, the
+   ``navbar-menu-nav.html`` template any anything that comes before it will
+   be **left-aligned**.
```
I am wondering here: does this signal that we should rather split it in three? And also have a `navbar_right` option for the content that is right-aligned?

jorisvandenbossche replied:
Hmm, right now the menu items and icons are in the same div, using a single class/alignment. So that's indeed a bit a bigger change to be able to split those.

jorisvandenbossche replied:
We still need to ensure that the middle and right are together in the collapsible div, so that the mobile responsiveness is correct. 

I did a quick attempt, see the commit I pushed. It seems to work the same as it does now (but didn't check if the alignment is still correct if you are using multiple custom templates)

jorisvandenbossche commented on pydata_sphinx_theme/_templates/navbar-menu-buttons.html:
```diff
@@ -0,0 +1,3 @@
+{%- block icon_links -%}
+{%- include "icon-links.html" with context -%}
```
Should we put the content of `icon-links.html` simply here?

choldgraf replied:
you mean and remove the `block`? I think the main benefit is if somebody wanted to override that section and add something of their own, do we want this to be possible?

drammock replied:
we currently override `block icon_links` in the MNE-Python site. So removing it would necessitate some changes for us. See https://github.com/mne-tools/mne-python/blob/main/doc/_templates/docs-navbar.html

jorisvandenbossche replied:
I didn't mean to remove the `block`, but just the content of the file could directly be put in this template file instaed of using the `include`?

Also, to keep to our terminology of "icon links", should we call this "navbar-menu-icons"? instead of "buttons"?

jorisvandenbossche replied:
@drammock I *think* that also with the current change, you will need to do a small update to your custom template. Because the `{%- block icon_links -%}` is now no longer defined in docs-navbar.html, but in this navbar-menu-buttons.html which gets inserted into docs-navbar.html.

(now, I don't know enough of sphinx / jinja to know in which order what happens)

drammock replied:
thanks for the heads-up.  I think we're pinned at the moment, but I'll make the necessary changes and unpin once this PR is merged.

jorisvandenbossche reviewed (commented):
Cool! Fully on board with the principles, thanks!

> Removes the configuration value to put the search in the sidebar or the navbar

Do you know if there is an easy way to deprecate this first?

choldgraf commented:
Thanks for the review @jorisvandenbossche - some quick comments above. For this one:

> Do you know if there is an easy way to deprecate this first?

Good point, what if we kept the flag, and for now if that flag exists it will manually update the nav configs to put the search template in the right place?

jorisvandenbossche commented:
> Good point, what if we kept the flag, and for now if that flag exists it will manually update the nav configs to put the search template in the right place? And raise a dep warning?

That sounds good to me

choldgraf commented:
@jorisvandenbossche ah nice - this almost gets there, but if you add extra items in addition to the menu, this happens:

![image](https://user-images.githubusercontent.com/1839645/113378047-e122dd00-932a-11eb-92fb-23c7be5be689.png)

any idea what's going wrong there?

choldgraf commented:
OK I think I've addressed all comments, added a deprecation, fixed the CSS styling, and got the tests working. Anything else to take care of?

@jorisvandenbossche I didn't change the `icon_links` stuff because those configuration values etc are already in there, so if we wanted to change it we'd have to add a deprecation for it as well. It seems reasonable to do, but maybe out of scope of this PR?

choldgraf commented:
@jorisvandenbossche what do you think about these two extra additions I can add to this PR:

1. Do the same thing for the `footer` as well as the `page toc` sections. So that instead of hard-coding them, we have a similar configuration value that provides a list of templates.
2. Make it possible for people to omit the `.html` in their configuration (e.g., `navbar_right: ["icon-links", "other-template"]`) via a simple script that just checks for whether an item has an extension, and if not, appends `.html`

Do those seem reasonable?

jorisvandenbossche commented:
Sounds good to me.

Just wondering, for the footer, I would think you will typically override the full footer? In which case you could also add a `footer.html` to your custom `_templates` ?

choldgraf commented:
OK @jorisvandenbossche - I've added in a `right_sidebar` and a `footer` section as well, and spun out the documentation for adding HTML templates to each section into its own page, which you can check out here:

https://pydata-sphinx-theme--355.org.readthedocs.build/en/355/user_guide/sections.html

jorisvandenbossche commented on pydata_sphinx_theme/_templates/last-updated.html:
```diff
@@ -0,0 +1,3 @@
+<p class="last-updated">
```
The original had a `if last_updated`, should that be kept?

choldgraf replied:
I was thinking that people could just provide this in the footer config if they wished, rather than using the Sphinx feature flag. Does that make sense?

jorisvandenbossche commented on src/scss/_navbar.scss:
```diff
@@ -7,6 +7,18 @@
+      float: left;
+    }
+
+    // Anything in navbar menu should be left-justified (e.g. links)
```
```suggestion
    // Anything in right part of the navbar should be right-justified (e.g. icons)
```

jorisvandenbossche reviewed (commented):
Cool, thanks a lot for the updates! Good idea for making a separate doc page about it

I added a bunch of comments. Large part of them are pedantic naming discussions, but since we let people override those based on the name, we should stick with the names once we introduce them, so a bit of bikeshedding can be useful then ;)

choldgraf commented:
Thanks for these comments @jorisvandenbossche - I implemented pretty much all of them (w/ one exception which I responded in-line above). Great points about naming things, I think "start" and "end" is much better. Let me know if there's anything else I should finish up (and maybe have a quick look that I didn't miss anything in the renaming)

choldgraf commented on docs/user_guide/sections.rst:
```diff
@@ -0,0 +1,127 @@
+configuration variable.
+
+For the left sidebar only, you can configure templates so that they only show
+up on certain pages. You do so via a configuration like so in ``conf.py``:
```
https://github.com/pydata/pydata-sphinx-theme/issues/376

jorisvandenbossche commented:
Thanks for the updates! I pushed a small fix to add again a bit of padding at the bottom of the footer (which was removed by removing the padding around the <p> tag in the footer)
